### ISSUE TITLE
Is TCN suitable for time series regression?

### ISSUE THREAD

**Issue by liuzf13** [locuslab/TCN#25]
Hello,

Thank you for your great paper and sharing!

I'm wondering how to use TCN to solve time series regression problems.  In my time series scenario, data for each moment contains multiple variables and each variable is a real number. For example, data for time step 0 is something like "vector_0 = <0.1, 0.2, 0.3, ...>", and I want to use historical k vectors to predict the next vector data.

I have developed a LSTM model for this question. The input shape of LSTM model is "batch_size, time_steps(k), input_size(length of each vector)", and the prediction result is the last value of LSTM. Then I could calculate the MSE loss and do backward. How can I use TCN to solve this problem? 

Best Regards

**Comment by jerrybai1995** [comment]
Hi, thank you for your interest in TCN.

We designed TCN as a drop-in replacement for LSTM, so as long as your sequence has a well-defined time step (as you described), then TCN can be used to solve this problem as well. Basically, in pytorch, anywhere you used to call `y = lstm(x, (h,c))`, you can now try to replace it with `y = tcn(x)`, with possibly some tuning.

Hope this helps!

**Comment by liuzf13** [comment]
Hi jerry, 

Thank you for your comment :-). 

I have figured it out by using the last value of the liner output layer(y1[:, :, -1]) as prediction value and added a sigmoid function to it for positive time series regression.  The MSE result of TCN and LSTM are similar, but TCN takes longer time for training. I'll try to adjust parameters to see if TCN can get better results.

By the way, is there any suggestions for the parameters? The length of each vector datais 529, and I use historical 10 vectors to predict the next vector.

**Comment by jerrybai1995** [comment]
Hi @liuzf13,

Do all input sequences have length 10, or are you only limiting the horizon to 10? Basically, the things I would consider when using a TCN is:

1. How large of an effective history do you need when making a prediction? (As an example, for MNIST, you probably need all 784 pixels in the sequence; whereas for word-level LM you probably need no more than 100 to build a good model.) Based on this size, you can choose the number of layers and dilation sizes. In your case, if you think 10 is good enough, I would suggest you try removing dilations and just apply a deeply stacked TCN.
2. How much data do you have? If the dataset is small, then regularization in TCN can be quite important. You can try to change the dropout rates, normalization scheme, or maybe even add auxiliary losses to the lower levels of the TCN.
3. Is your 529-dim vector relatively sparse? Did you try to pass the sequence through a dense layer to first compress these vectors (these may help the generalization of the model, too).

On long sequences, I think TCN is typically more efficient, but on a short sequence (e.g. length-10), it is not a surprise that LSTM can be as fast as TCN, given the optimization of RNNs in cuDNN.

**Comment by liuzf13** [comment]
Hi @jerrybai1995,

Thank you for your suggestions. The size of my data set is around 3500 (I sampled 529-dim data every 15 minutes for about 5 weeks), and I use min-max normalization for every dimension. The sequence length is just a parameter and I'll try to adjust it.
